We review the first two parts of the course for the final exam.

Below are the answers to the review problems for the first part of the course.

# 1. Our First Steps

### question 1.

 Explain the difference between ``RDF`` and ``RR``.  Illustrate with an example.

### answer to question 1

 ``RDF`` corresponds to the hardware doubles,
whereas ``RR`` has a default precision of 53 bits.

In [1]:
reset()

In [2]:
print(RDF)
print(RR)

Real Double Field
Real Field with 53 bits of precision


In [3]:
a = RDF(pi)
print(a, 'has type', type(a))
b = RR(pi)
print(b, 'has type', type(b))

3.141592653589793 has type <class 'sage.rings.real_double_element_gsl.RealDoubleElement_gsl'>
3.14159265358979 has type <class 'sage.rings.real_mpfr.RealNumber'>


### question 2

Give a list of ten rational approximations for ``sqrt(3)``, accurate with 2, 3, up to 11 decimal places.

### answer to question 2

In [4]:
reset()

In [5]:
x = sqrt(3)
[QQ(x.n(digits=k)) for k in range(2,12)]

[26/15,
 265/153,
 362/209,
 989/571,
 8733/5042,
 18817/10864,
 70226/40545,
 191861/110771,
 716035/413403,
 1694157/978122]

### question 3

Consider $Z_{31}$.  What is the multiplicative inverse of 7?
   
Show that $15 x^5 + 4 x^4 + 23 x^3 + 26 x^2 + 6 x + 1$ is irreducible over $Z_{31}$.

### answer to question 3

In [6]:
reset()

In [7]:
Z31 = GF(31)
Z31.multiplication_table()

 *  aa ab ac ad ae af ag ah ai aj ak al am an ao ap aq ar as at au av aw ax ay az ba bb bc bd be
  +---------------------------------------------------------------------------------------------
aa| aa aa aa aa aa aa aa aa aa aa aa aa aa aa aa aa aa aa aa aa aa aa aa aa aa aa aa aa aa aa aa
ab| aa ab ac ad ae af ag ah ai aj ak al am an ao ap aq ar as at au av aw ax ay az ba bb bc bd be
ac| aa ac ae ag ai ak am ao aq as au aw ay ba bc be ab ad af ah aj al an ap ar at av ax az bb bd
ad| aa ad ag aj am ap as av ay bb be ac af ai al ao ar au ax ba bd ab ae ah ak an aq at aw az bc
ae| aa ae ai am aq au ay bc ab af aj an ar av az bd ac ag ak ao as aw ba be ad ah al ap at ax bb
af| aa af ak ap au az be ae aj ao at ay bd ad ai an as ax bc ac ah am ar aw bb ab ag al aq av ba
ag| aa ag am as ay be af al ar ax bd ae ak aq aw bc ad aj ap av bb ac ai ao au ba ab ah an at az
ah| aa ah ao av bc ae al as az ab ai ap aw bd af am at ba ac aj aq ax be ag an au bb ad ak ar ay
ai| aa ai aq ay ab aj ar az ac

In [8]:
i7 = Z31(1/7)
i7

9

We see that the multiplicative inverse of 7 modulo 31 is 9.  Let us verify this:

In [9]:
Z31(7*i7)

1

Indeed, 9 is the inverse of 7 modulo 31.

Let us now solve the second part of the question.

In [10]:
x = polygen(Z31)
p = 15*x^5 + 4*x^4 + 23*x^3 + 26*x^2 + 6*x + 1
p

15*x^5 + 4*x^4 + 23*x^3 + 26*x^2 + 6*x + 1

In [11]:
type(p)

<class 'sage.rings.polynomial.polynomial_zmod_flint.Polynomial_zmod_flint'>

The ``type(p)`` verifies that ``p`` is not a general symbolic expression, but a particular type of polynomial.

In [12]:
p.factor()

(15) * (x^5 + 23*x^4 + 16*x^3 + 10*x^2 + 19*x + 29)

In [13]:
p.is_irreducible()

True

Both ``p.factor()`` and ``p.is_irreducible()`` show that the polynomial is irreducible.

Instead of ``polygen(GF(31))``, we can also use ``PolynomialRing``,
which gives the same type of polynomial.

In [14]:
reset()

In [15]:
R.<x> = PolynomialRing(GF(31))
p = 15*x^5 + 4*x^4 + 23*x^3 + 26*x^2 + 6*x + 1
print(p, 'has type', type(p))

15*x^5 + 4*x^4 + 23*x^3 + 26*x^2 + 6*x + 1 has type <class 'sage.rings.polynomial.polynomial_zmod_flint.Polynomial_zmod_flint'>


In [16]:
p.is_irreducible()

True

### question 4

Expand ``exp(I*2*Pi/k)`` as ``cos(2*Pi/k) + I*sin(2*Pi/k)``.

### answer to question 4

In [17]:
reset()

In [18]:
k = SR.symbol('k', domain='real')
t = exp(I*2*pi/k)
t

e^(2*I*pi/k)

In [19]:
t.real() + I*t.imag()

cos(2*pi/k) + I*sin(2*pi/k)

### question 5

Consider ``q = cos(x^3 - 1) + 3*sin(y) - z^7``.  Draw the expression tree of ``q``.

### answer to question 5

In [20]:
reset()

In [21]:
x, y, z = var('x,y,z')
q = cos(x^3 - 1) + 3*sin(y) - z^7
q

-z^7 + cos(x^3 - 1) + 3*sin(y)

In [22]:
q.operands()

[-z^7, cos(x^3 - 1), 3*sin(y)]

In [23]:
q.operator()

<function add_vararg at 0x7f0fc7082830>

In [24]:
(q0, q1, q2) = q.operands()
print('q0 :', q0.operator(), q0.operands())
print('q1 :', q1.operator(), q1.operands())
print('q2 :', q2.operator(), q2.operands())

q0 : <function mul_vararg at 0x7f0fc70828c0> [z^7, -1]
q1 : cos [x^3 - 1]
q2 : <function mul_vararg at 0x7f0fc70828c0> [sin(y), 3]


In [25]:
q00 = q0.operands()[0]
print('q00 :', q00.operator(), q00.operands())

q00 : <built-in function pow> [z, 7]


Now we have enough information to draw the expression tree.

Let us start representing all leaves.

In [26]:
Lx = LabelledOrderedTree([], label='x')
Ly = LabelledOrderedTree([], label='y')
Lz = LabelledOrderedTree([], label='z')
L7 = LabelledOrderedTree([], label='7')
Lm1 = LabelledOrderedTree([], label='-1')
L3 = LabelledOrderedTree([], label='3')

We start with the ``- z^7``.

In [27]:
Lz7 = LabelledOrderedTree([Lz,L7], label='pow')
Lop1 = LabelledOrderedTree([Lz7,Lm1], label='mul')
ascii_art(Lop1)

<repr(<sage.typeset.ascii_art.AsciiArt at 0x7f0fc36c4700>) failed: OSError: [Errno 25] Inappropriate ioctl for device>

The second operand is ``cos(x^3-1)``.

In [28]:
Lx3 = LabelledOrderedTree([Lx,L3], label='pow')
Lx3m1 = LabelledOrderedTree([Lx3,Lm1], label='add')
Lop2 = LabelledOrderedTree([Lx3m1], label='cos')
ascii_art(Lop2)

<repr(<sage.typeset.ascii_art.AsciiArt at 0x7f0fc36c5870>) failed: OSError: [Errno 25] Inappropriate ioctl for device>

Then the third operand is ``3*sin(y)``

In [29]:
Lsin = LabelledOrderedTree([Ly], label='sin')
Lop3 = LabelledOrderedTree([L3,Lsin], label='mul')
ascii_art(Lop3)

<repr(<sage.typeset.ascii_art.AsciiArt at 0x7f0fc36c4340>) failed: OSError: [Errno 25] Inappropriate ioctl for device>

Now we can assemble the tree.

In [30]:
tree = LabelledOrderedTree([Lop1,Lop2,Lop3], label='add')
ascii_art(tree)

<repr(<sage.typeset.ascii_art.AsciiArt at 0x7f0fc36c6500>) failed: OSError: [Errno 25] Inappropriate ioctl for device>

### question 6

Save a random matrix to file, destroy its reference,
and load then the saved matrix back into SageMath.

### answer question 6

In [31]:
reset()

In [32]:
M = Matrix(3,4,lambda i,j: RR.random_element())
M

[  0.621660021942880   0.649467962756092  -0.554103848766790  0.0227078641543150]
[  0.404637308293002  -0.722876671193864  -0.266150461564519 -0.0663804221933906]
[ -0.141812483491682  -0.694540150031590  -0.167627160357438   0.693178157317415]

In [33]:
M.save('ourmatrix')

In [34]:
M = var('M')
M

M

In [35]:
M = load('ourmatrix.sobj')
M

[  0.621660021942880   0.649467962756092  -0.554103848766790  0.0227078641543150]
[  0.404637308293002  -0.722876671193864  -0.266150461564519 -0.0663804221933906]
[ -0.141812483491682  -0.694540150031590  -0.167627160357438   0.693178157317415]

### question 7

Consider ``p = 79*x^298+56*x^205+49*x^164+63*x^121+57*x^119−59*x^42``.

1. Generate optimized code to evaluate ``p``.
  
2. How many operations are needed to evaluate ``p``?
   
3. Compare with the cost of the direct evaluation of ``p``.

### answer to question 7

In [36]:
reset()

In [37]:
x = var('x')
p = 79*x^298  + 56*x^205 + 49*x^164 + 63*x^121 + 57*x^119 - 59*x^42
p

79*x^298 + 56*x^205 + 49*x^164 + 63*x^121 + 57*x^119 - 59*x^42

In [38]:
q = fast_callable(p,vars=['x'])
q.op_list()

[('load_arg', 0),
 ('ipow', 298),
 ('load_const', 79),
 'mul',
 ('load_arg', 0),
 ('ipow', 205),
 ('load_const', 56),
 'mul',
 'add',
 ('load_arg', 0),
 ('ipow', 164),
 ('load_const', 49),
 'mul',
 'add',
 ('load_arg', 0),
 ('ipow', 121),
 ('load_const', 63),
 'mul',
 'add',
 ('load_arg', 0),
 ('ipow', 119),
 ('load_const', 57),
 'mul',
 'add',
 ('load_arg', 0),
 ('ipow', 42),
 ('load_const', -59),
 'mul',
 'add',
 'return']

In [39]:
print('number of operations :', len(q.op_list()))

number of operations : 30


In the ``fast_callable`` evaluation, we have six exponentiations (``ipow``),
six multiplications with a coefficient, and five additions.
In the direct evaluation, without ``ipow``, the number of multiplications is
the sum of the degrees of the operands of p.

In [40]:
pdegs = [t.degree(x) for t in p.operands()]
pdegs

[298, 205, 164, 121, 119, 42]

In [41]:
print('number of multiplications:', sum(pdegs))

number of multiplications: 949


### question 8

Consider the right hand side of

$$
   \int_1^2 \frac{1}{x} dx \approx \frac{1}{2n} \left( 1 + \frac{1}{2} \right)
                                  + \frac{1}{n} \sum_{k=1}^n \left( \frac{1}{1 + k/n} \right).
$$

1. Write a Python function ``P`` to define the evaluation of the right hand side of the above sum,
   for any value for $n$ which is the input parameter to the function.
   
   Compare the value of ``P(10000)`` with the exact value of the integral.
   
   Time the execution of the function for $n = 10,000$.
   
   
2. Use vectorization to define a more efficient function ``F`` which computes the same sum as ``P``
   and which also takes $n$ as the input parameter.

   Compare the value of ``F(10000)`` with the exact value of the integral.
   
   Time the execution of the function for $n = 10,000$.
   
Compare the result of the timings.

### answer to question 8

In [42]:
reset()

In [43]:
P = lambda n: (1.0 + 0.5)/(2*n) + sum(1/(1+k/n) for k in range(1, n))/n 

In [44]:
p10000 = P(10000)
p10000

0.693147181184945

In [45]:
exact = integral(1/x, (x, 1, 2), hold=True).n()
exact

0.6931471805599452

In [46]:
p10000 - exact

6.25000162735034e-10

We see that the error of the approximate sum is about $6.25 \times 10^{-10}$ which is accurate enough.

In [47]:
timeit('P(10000)')

25 loops, best of 3: 16.9 ms per loop

Now we import ``numpy`` to apply the vectorization.

In [48]:
import numpy as np

In [49]:
def F(n):
    """
    Applies vectorization to the trapezoidal rule
    with n function evaluations.
    """
    pts = np.linspace(1.0 + 1.0/n, 2.0 - 1.0/n, n-1)
    return ((1.0 + 0.5)/(2*n) + sum(1/pts)/n)

In [50]:
f10000 = F(10000)
f10000

0.6931471811849453

In [51]:
f10000 - exact

6.250001627350343e-10

We see the same error as before.

In [52]:
timeit('F(10000)')

625 loops, best of 3: 42.8 μs per loop

We observe that by vectorization, the time has dropped from 52 milliseconds to 151 microseconds.

# 2. Polynomials and Rational Expressions

### question 9

Draw the expression tree of ``x*y*(x-y)``.

### answer to question 9

In [53]:
reset()

In [54]:
x, y = var('x,y')
p = x*y*(x-y)
p

(x - y)*x*y

In [55]:
p.operator()

<function mul_vararg at 0x7f0fc70828c0>

In [56]:
p.operands()

[x - y, x, y]

In [57]:
p0 = p.operands()[0]
print('p0 :', p0, p0.operator(), p0.operands())

p0 : x - y <function add_vararg at 0x7f0fc7082830> [x, -y]


In [58]:
p1 = p0.operands()[1]
print('p1 :', p1, p1.operator(), p1.operands())

p1 : -y <function mul_vararg at 0x7f0fc70828c0> [y, -1]


We now have enough information to draw the expression tree.

To represent the tree via LabelledOrderedTree,
we can start defining the leaves: ``x``, ``y``, and ``-1``.

In [59]:
Lx = LabelledOrderedTree([], label='x')
Ly = LabelledOrderedTree([], label='y')
Lm1 = LabelledOrderedTree([], label='-1')
# The x - y is represented as x + (-1)*y
Lmy = LabelledOrderedTree([Ly,Lm1],label='mul')
Lop1 = LabelledOrderedTree([Lx,Lmy],label='add')
ascii_art(Lop1)

<repr(<sage.typeset.ascii_art.AsciiArt at 0x7f0fb917d420>) failed: OSError: [Errno 25] Inappropriate ioctl for device>

Then the expression tree has Lop1 as the first child,
the other two children are ``Lx`` and ``Ly``.

In [60]:
tree = LabelledOrderedTree([Lop1, Lx, Ly], label='mul')
ascii_art(tree)

<repr(<sage.typeset.ascii_art.AsciiArt at 0x7f0fb917c2b0>) failed: OSError: [Errno 25] Inappropriate ioctl for device>

### question 10

Consider $p = x^3 - x - 2$ and give all Sage commands for an exact,
numeric, and symbolic factorization.

In particular, answer the following questions.
   
1. How can you write $p$ as an *exact* product of
   linear factors, with *exact* complex numbers?
      
2. Compute a *numerical* factorization
   of $p$ over the complex numbers.
      
3. Define a *symbolic* (i.e.: formal) factorization of $p$,
   declaring sufficiently many roots.
   
4. Compute a *symbolic-numeric* factorization of $p$.

### answer to question 10

In [61]:
reset()

In [62]:
x = var('x')
p = x^3 - x - 2
p

x^3 - x - 2

In [63]:
r = p.roots()
r

[(-1/2*(1/9*sqrt(26)*sqrt(3) + 1)^(1/3)*(I*sqrt(3) + 1) - 1/6*(-I*sqrt(3) + 1)/(1/9*sqrt(26)*sqrt(3) + 1)^(1/3),
  1),
 (-1/2*(1/9*sqrt(26)*sqrt(3) + 1)^(1/3)*(-I*sqrt(3) + 1) - 1/6*(I*sqrt(3) + 1)/(1/9*sqrt(26)*sqrt(3) + 1)^(1/3),
  1),
 ((1/9*sqrt(26)*sqrt(3) + 1)^(1/3) + 1/3/(1/9*sqrt(26)*sqrt(3) + 1)^(1/3), 1)]

We can assemble the exact factorization from the exact roots.

In [64]:
prod([x - root[0] for root in r])

1/108*(3*(1/9*sqrt(26)*sqrt(3) + 1)^(1/3)*(I*sqrt(3) + 1) + 6*x + (-I*sqrt(3) + 1)/(1/9*sqrt(26)*sqrt(3) + 1)^(1/3))*(3*(1/9*sqrt(26)*sqrt(3) + 1)^(1/3)*(-I*sqrt(3) + 1) + 6*x + (I*sqrt(3) + 1)/(1/9*sqrt(26)*sqrt(3) + 1)^(1/3))*(3*x - 3*(1/9*sqrt(26)*sqrt(3) + 1)^(1/3) - 1/(1/9*sqrt(26)*sqrt(3) + 1)^(1/3))

The numerical factorization is over the complex number field.

In [65]:
q = CC['x'](p)
q.factor()

(x - 1.52137970680457) * (x + 0.760689853402284 - 0.857873626595179*I) * (x + 0.760689853402284 + 0.857873626595179*I)

Let us now do the symbolic factorization.

In [66]:
K.<a> = NumberField(p) 
K

Number Field in a with defining polynomial x^3 - x - 2

In [67]:
kp = K['x'](p)
kp.factor()

(x - a) * (x^2 + a*x + a^2 - 1)

In [68]:
kpb = list(kp.factor())[1]
L.<b> = NumberField(kpb[0])
L

Number Field in b with defining polynomial x^2 + a*x + a^2 - 1 over its base field

In [69]:
lkp = L['x'](kp)
lkp.factor()

(x + b + a) * (x - b) * (x - a)

The symbolic-numeric factorization works over the field ``QQbar``, extending the complex number with complex intervals.

In [70]:
factor(QQbar[x](p))

(x - 1.521379706804568?) * (x + 0.7606898534022837? - 0.8578736265951787?*I) * (x + 0.7606898534022837? + 0.8578736265951787?*I)

### question 11

Give all the sage commands to convert ``(x-y)*(x+y)`` into ``(x+y)*x = (x+y)*y``.

### answer to question 11

In [71]:
reset()

In [72]:
x,y,z = var('x,y,z')
p = (x-y)*(x+y)
p

(x + y)*(x - y)

In [73]:
q = p.subs({x+y:z})
q

(x - y)*z

In [74]:
r = q.expand()
r.subs(z=x+y)

(x + y)*x - (x + y)*y

### question 12

Consider $r = (79*x^5 + 56*x^4 + 49*x^3 + 63*x^2 + 57*x - 59)/(45*x^5 - 8*x^4 - 93*x^2 + 43*x - 62)$.

1. How many operations does it take to evaluate $r$ with a partial fraction decomposition? 

2. Compare this number of operations with the Horner forms of numerator and denominator for r.

### answer to question 12

In [75]:
reset()

In [76]:
R.<x> = PolynomialRing(CC)
r = (79*x^5 + 56*x^4 + 49*x^3 + 63*x^2 + 57*x - 59)/(45*x^5 - 8*x^4 - 93*x^2 + 43*x - 62)
show(r)

(79.0000000000000*x^5 + 56.0000000000000*x^4 + 49.0000000000000*x^3 + 63.0000000000000*x^2 + 57.0000000000000*x - 59.0000000000000)/(45.0000000000000*x^5 - 8.00000000000000*x^4 - 93.0000000000000*x^2 + 43.0000000000000*x - 62.0000000000000)

In [77]:
q = r.partial_fraction_decomposition()
q

(1.75555555555556,
 [1.73033923026938/(x - 1.34613302859424),
  (-0.209546357895326 - 0.317866843685468*I)/(x - 0.243395948193229 - 0.710544938063163*I),
  (-0.209546357895326 + 0.317866843685468*I)/(x - 0.243395948193229 + 0.710544938063163*I),
  (0.122648347698910 + 0.522394270600454*I)/(x + 0.827573573601461 - 1.06277001016849*I),
  (0.122648347698910 - 0.522394270600454*I)/(x + 0.827573573601461 + 1.06277001016849*I)])

If we would use the Horner form to evaluate numerator and denominator,
then we need 5 multiplications and 5 additions, with one division.
The partial fraction decomposition gives 5 divisions and 5 (complex) additions.

### question 13

Explain why normal forms are so important to symbolic computation. 

What can we do if a normal form is too expensive to compute? Illustrate with a good example.

### answer to question 13

With a normal form we can decide whether two expressions are equal to each other.
The normalization for rational expression is the removal of common factors in the numerator
and the denominator.  This may lead to expression swell, as the example below shows.
We can apply a probability one numerical evaluation test.
We generate a random number and evaluate the expression to see if the expression is zero or not.

In [78]:
reset()

In [79]:
x = var('x')
p = (x^100 - 1)/(x - 1)
p

(x^100 - 1)/(x - 1)

In [80]:
p.numerator()

x^99 + x^98 + x^97 + x^96 + x^95 + x^94 + x^93 + x^92 + x^91 + x^90 + x^89 + x^88 + x^87 + x^86 + x^85 + x^84 + x^83 + x^82 + x^81 + x^80 + x^79 + x^78 + x^77 + x^76 + x^75 + x^74 + x^73 + x^72 + x^71 + x^70 + x^69 + x^68 + x^67 + x^66 + x^65 + x^64 + x^63 + x^62 + x^61 + x^60 + x^59 + x^58 + x^57 + x^56 + x^55 + x^54 + x^53 + x^52 + x^51 + x^50 + x^49 + x^48 + x^47 + x^46 + x^45 + x^44 + x^43 + x^42 + x^41 + x^40 + x^39 + x^38 + x^37 + x^36 + x^35 + x^34 + x^33 + x^32 + x^31 + x^30 + x^29 + x^28 + x^27 + x^26 + x^25 + x^24 + x^23 + x^22 + x^21 + x^20 + x^19 + x^18 + x^17 + x^16 + x^15 + x^14 + x^13 + x^12 + x^11 + x^10 + x^9 + x^8 + x^7 + x^6 + x^5 + x^4 + x^3 + x^2 + x + 1

In [81]:
p(x=RR.random_element())

0.685474316298453

We see that, when evaluated at a random element, the nonzero value shows that the expression is not zero.